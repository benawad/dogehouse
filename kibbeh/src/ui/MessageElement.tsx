import React from "react";
import { formatDistance, fromUnixTime } from "date-fns";

import { SingleUser } from "./UserAvatar";

export interface MessageElementProps {
  user: {
    username: string;
    avatar: string;
    isOnline: boolean;
  };
  msg: {
    ts: number;
    text: string;
  };
}

interface MessageDateProps {
  ts: number;
}

const MessageDate: React.FC<MessageDateProps> = ({ ts }) => (
  <span
    className="text-primary-300 text-sm font-medium inline-block truncate"
    style={{
      lineHeight: "22px",
    }}
  >
    {formatDistance(fromUnixTime(ts), new Date())} ago
  </span>
);

export const MessageElement: React.FC<MessageElementProps> = ({
  user,
  msg,
}) => {
  return (
    <div
      className="items-center w-full px-4 md:bg-primary-800 md:border-b md:border-primary-600 cursor-pointer hover:bg-primary-700 bg-primary-900"
      data-testid="msg-element"
    >
      <div className="flex mr-3">
        <SingleUser size="sm" isOnline={user.isOnline} src={user.avatar} />
      </div>
      <div
        className="flex-col py-3 border-b border-primary-600 md:border-none md:py-0"
        style={{
          width: "calc(100% - 50px)",
        }}
      >
        <div className="flex justify-between">
          <span
            className="text-button font-bold inline-block truncate mr-1"
            style={{
              lineHeight: "22px",
            }}
          >
            {user.username}
          </span>
          <MessageDate ts={msg.ts} />
        </div>
        <div
          className="block text-sm text-primary-300 font-medium truncate w-9/12"
          style={{
            lineHeight: "22px",
          }}
        >
          {msg.text}
        </div>
      </div>
    </div>
  );
};
