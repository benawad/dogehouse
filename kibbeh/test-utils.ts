import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import "mutationobserver-shim";

const customRender = (ui: any, options = {}) =>
  render(ui, {
    ...options,
  });

export * from "@testing-library/react";

export { customRender as render };
