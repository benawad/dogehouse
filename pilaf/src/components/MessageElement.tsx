import { formatDistance, fromUnixTime } from "date-fns";
import React, { useState } from "react";
import {
  ImageSourcePropType,
  StyleSheet,
  Text,
  TextStyle,
  TouchableOpacity,
  View,
  ViewProps,
} from "react-native";
import { colors, paragraphBold, small } from "../constants/dogeStyle";
import { SingleUserAvatar } from "./avatars/SingleUserAvatar";

export type MessageElementProps = ViewProps & {
  user: {
    username: string;
    avatar: ImageSourcePropType;
    isOnline: boolean;
  };
  msg: {
    ts: number;
    text: string;
  };
};

interface MessageDateProps {
  ts: number;
  style?: TextStyle;
}

const MessageDate: React.FC<MessageDateProps> = ({ ts, style }) => (
  <Text style={style}>{formatDistance(fromUnixTime(ts), new Date())} ago</Text>
);

export const MessageElement: React.FC<MessageElementProps> = ({
  style,
  user,
  msg,
}) => {
  const [expanded, setExpanded] = useState(false);
  return (
    <TouchableOpacity
      onPress={() => setExpanded(!expanded)}
      activeOpacity={0.8}
    >
      <View style={[styles.container, style]}>
        <SingleUserAvatar
          size={"sm"}
          isOnline={user.isOnline}
          src={user.avatar}
          style={{ marginBottom: 20 }}
        />
        <View style={styles.middleContainer}>
          <Text style={styles.textUserName}>{user.username}</Text>
          <Text style={styles.textMessage} numberOfLines={1}>
            {msg.text}
          </Text>
        </View>
        <View style={styles.dateContainer}>
          <MessageDate style={styles.textDate} ts={msg.ts} />
        </View>
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    alignItems: "center",
  },
  middleContainer: {
    flex: 1,
    borderBottomWidth: 1,
    borderBottomColor: colors.primary700,
    paddingBottom: 20,
    paddingLeft: 0,
    marginLeft: 11,
    justifyContent: "space-between",
    overflow: "hidden",
  },
  dateContainer: {
    height: "100%",
    flexDirection: "row",
    alignItems: "flex-start",

    justifyContent: "center",
    borderBottomWidth: 1,
    borderBottomColor: colors.primary700,
  },
  textUserName: {
    ...paragraphBold,
  },
  textDate: {
    ...small,
    color: colors.primary300,
  },
  textMessage: {
    ...small,
    color: colors.primary300,
  },
});
