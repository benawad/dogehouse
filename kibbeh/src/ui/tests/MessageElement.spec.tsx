import React from "react";

import { render } from "../../../test-utils";
import { MessageElement } from "../MessageElement";

describe("MessageElement", () => {
  it("should render correctly", () => {
    const { getByTestId } = render(
      <MessageElement
        msg={{ text: "some text", ts: 123 }}
        user={{ avatar: "", isOnline: true, username: "someUserNAme" }}
      />
    );
    const component = getByTestId("msg-element");

    expect(component).toBeVisible();
  });

  it("should match snapshot", () => {
    const { getByTestId } = render(
      <MessageElement
        msg={{ text: "some text", ts: 123 }}
        user={{ avatar: "", isOnline: true, username: "someUserNAme" }}
      />
    );
    const component = getByTestId("msg-element");
    expect(component).toMatchSnapshot();
  });
});
