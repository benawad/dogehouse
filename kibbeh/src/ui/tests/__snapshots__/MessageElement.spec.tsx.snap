// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageElement should match snapshot 1`] = `
<div
  class="items-center w-full px-4 md:bg-primary-800 md:border-b md:border-primary-600 cursor-pointer hover:bg-primary-700 bg-primary-900"
  data-testid="msg-element"
>
  <div
    class="flex mr-3"
  >
    <div
      class="relative inline-block "
      data-testid="single-user-avatar"
      style="width: 40px; height: 40px;"
    >
      <img
        alt="your-avatar"
        class="rounded-full w-full h-full object-cover "
        src=""
      />
      <span
        class="rounded-full absolute box-content bg-accent border-primary-800"
        data-testid="online-indictor"
        style="width: 8px; height: 8px; right: 2px; bottom: -2px; border-width: 2px;"
      />
    </div>
  </div>
  <div
    class="flex-col py-3 border-b border-primary-600 md:border-none md:py-0"
    style="width: calc(100% - 50px);"
  >
    <div
      class="flex justify-between"
    >
      <span
        class="text-button font-bold inline-block truncate mr-1"
        style="line-height: 22px;"
      >
        someUserNAme
      </span>
      <span
        class="text-primary-300 text-sm font-medium inline-block truncate"
        style="line-height: 22px;"
      >
        over 51 years
         ago
      </span>
    </div>
    <div
      class="block text-sm text-primary-300 font-medium truncate w-9/12"
      style="line-height: 22px;"
    >
      some text
    </div>
  </div>
</div>
`;
